File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.js

var groupThePeople = function(arr) {
       let ans = [];
       let group = {};
       for(let i =0; i< arr.length; i++){
           const size = arr[i];
           if(!group[size]){
               group[size] = [];
           }
           group[size].push(i);
           if(group[size].length === size){
               ans.push([...group[size]]);
               group[size] = [];
           }

       }
    return ans;    
};

File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.cpp
class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& arr) {
        vector<vector<int>> ans;
        unordered_map<int, vector<int>> mp;
        int n = arr.size();
        for(int i=0; i<n;i++){
            int size = arr[i];
            mp[size].push_back(i);
            if(mp[size].size() == size){
                ans.push_back(mp[size]);
                mp[size].clear();
            }
        }
        return ans;
    }
};